lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,230 @@
+        buildinfo = yaml.load(content)
+    return buildinfo
+
+
+def write_buildinfo_file(spec):
+    """
+    Create a cache file containing information
+    required for the relocation
+    """
+    text_to_relocate = []
+    binary_to_relocate = []
+    blacklist = (".spack", "man")
+    for root, dirs, files in os.walk(spec.prefix, topdown=True):
+        dirs[:] = [d for d in dirs if d not in blacklist]
+        for filename in files:
+            path_name = os.path.join(root, filename)
+            filetype = spack.relocate.get_filetype(path_name)
+            if spack.relocate.needs_binary_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                binary_to_relocate.append(rel_path_name)
+            elif spack.relocate.needs_text_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                text_to_relocate.append(rel_path_name)
+
+    # Create buildinfo data and write it to disk
+    buildinfo = {}
+    buildinfo['buildpath'] = spack.store.layout.root
+    buildinfo['relocate_textfiles'] = text_to_relocate
+    buildinfo['relocate_binaries'] = binary_to_relocate
+    filename = buildinfo_file_name(spec)
+    with open(filename, 'w') as outfile:
+        outfile.write(yaml.dump(buildinfo, default_flow_style=True))
+
+
+def tarball_directory_name(spec):
+    """
+    Return name of the tarball directory according to the convention
+    <os>-<architecture>/<compiler>/<package>/
+    """
+    return "%s/%s/%s" % (get_full_system_from_platform(),
+                         str(spec.compiler).replace("@", "-"),
+                         spec.name)
+
+
+def tarball_name(spec):
+    """
+    Return the name of the tarfile according to the convention
+    <os>-<architecture>-<package>-<dag_hash>.tar.gz
+    """
+    return "%s-%s-%s-%s.tar.gz" % (get_full_system_from_platform(),
+                                   spec.name,
+                                   spec.version,
+                                   spec.dag_hash())
+
+
+def tarball_path_name(spec):
+    """
+    Return the full path+name for a given spec according to the convention
+    <tarball_directory_name>/<tarball_name>
+    """
+    return os.path.join(tarball_directory_name(spec),
+                        tarball_name(spec))
+
+
+def build_tarball(spec, outdir, force=False):
+    """
+    Build a tarball from given spec and put it into the directory structure
+    used at the mirror (following <tarball_directory_name>).
+    """
+    tarfile_dir = os.path.join(outdir, tarball_directory_name(spec))
+    tarfile = os.path.join(outdir, tarball_path_name(spec))
+    if os.path.exists(tarfile):
+        if force:
+            os.remove(tarfile)
+        else:
+            tty.warn("file exists, use -f to force overwrite: %s" % tarfile)
+            return
+    if not os.path.exists(tarfile_dir):
+        mkdirp(tarfile_dir)
+
+    tar = which('tar', required=True)
```

Comment:
I believe @mathstuf is referring to https://docs.python.org/2/library/tarfile.html